Python requirements update

kbrock opened the pull request:
Overview
========

Our current python packages are not compatible with the current ansible version installed (and/or the current python version).

CP4AIOPS-5619 / https://github.com/ManageIQ/manageiq-rpm_build/pull/495 was just focused on vmware and aws.

This provides all requirements.

Issues
======

In a container, not able to build gssapi (for pyspnego[kerberos], for ms/AD)
This downloads many versions of `gssapi` but keeps failing.
This may be because I'm doing this inside docker instead of at install time.

Reference
=====

Dependencies brought up to date for:

- amazon/aws
- ansible/netcommon
- ansible/utils
- awx/awx
- azure/azcollection
- cisco/aci
- cisco/asa
- cisco/dnac
- cisco/intersight
- cisco/ios
- cisco/iosxr
- cisco/ise
- cisco/mso
- cisco/nxos
- cisco/ucs
- community/aws
- community/okd
- community/vmware
- google/cloud
- kubernetes/core
- microsoft/ad
- openstack/cloud
- ovirt/ovirt
- theforeman/foreman
- vmware/vmware_rest

kbrock commented on config/requirements.txt:
```diff
@@ -1,45 +1,66 @@
-azure-mgmt-storage==3.1.0
-azure-mgmt-trafficmanager==0.50.0
-azure-mgmt-web==0.41.0
-azure-storage==0.35.1
```
this was dropped (think replaced by azure-storage-blob

Don't see `azure-storage` listed in https://learn.microsoft.com/en-us/python/api/overview/azure/storage?view=azure-python

Fryguy commented on config/requirements.txt:
```diff
@@ -1,51 +1,100 @@
-psutil==5.6.6 # Match the version installed directly into the venv
-pykerberos==1.2.1
+ciscoisesdk>=2.0.8 # cisco/ise
+cryptography>=36.0.0 # cisco/intersight, cisco/aci, microsoft/ad
```
This matches rpm as well which is `python3-cryptography-36.0.1-4.el9.x86_64`, so this is good.

Fryguy commented on config/requirements.txt:
```diff
@@ -1,51 +1,100 @@
-pexpect==4.6.0
-psutil==5.6.6 # Match the version installed directly into the venv
-pykerberos==1.2.1
+ciscoisesdk>=2.0.8 # cisco/ise
```
This makes me think we should drop support for cisco/ise and the ones we don't necessarily support since it's introducing new packages. cc @agrare

Fryguy commented on config/requirements.txt:
```diff
@@ -1,51 +1,100 @@
+google-auth==1.6.2 # google/cloud
+google-cloud-storage # google/cloud
+grpcio # ansible/netcommon, cisco/iosxr
+ipaddress==1.0.23 # cisco/iosxr
```
I would love to drop ipaddress because it has had CVEs in the recent past, and we don't necessarily support cisco/iosxr

Fryguy commented:
This PR is fantastic - thanks for figuring this all out

Fryguy commented:
@kbrock Recapping our private convo with @agrare , let's drop all of the ciscos from this list except for cisco/intersight.

Fryguy commented:
> This downloads many versions of gssapi but keeps failing.

What is the error you're getting?  We might need to install a system package

kbrock commented:
@Fryguy It looks like we are missing `krb5-devel`, but we install that.

```
pip install 'pyspnego[kerberos]>=0.8.0'
Requirement already satisfied: pyspnego>=0.8.0 in /var/lib/manageiq/venv/lib/python3.9/site-packages (from pyspnego[kerberos]>=0.8.0) (0.11.1)
Requirement already satisfied: cryptography in /var/lib/manageiq/venv/lib64/python3.9/site-packages (from pyspnego>=0.8.0->pyspnego[kerberos]>=0.8.0) (43.0.1)
Collecting gssapi>=1.6.0 (from pyspnego[kerberos]>=0.8.0)
  Using cached gssapi-1.8.3.tar.gz (94 kB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... error
  error: subprocess-exited-with-error

  × Getting requirements to build wheel did not run successfully.
  │ exit code: 1
  ╰─> [21 lines of output]
      /bin/sh: line 1: krb5-config: command not found
```

Fryguy commented on bin/requirements.rb:
```diff
@@ -0,0 +1,187 @@
+#
+# 2. Resolve conflicts and determine if new one is correct
+#
+#     diff {,new-}requirements.txt
```
typo

```suggestion
#     diff {,new_}requirements.txt
```

Fryguy commented on bin/requirements.rb:
```diff
@@ -0,0 +1,187 @@
+      # the list of requirements-files can have items commented out - ignore those
+
+      mod = module_name_from_filename(fn)
+      IO.read(fn).split("\n").each do |line|
```
Doesnt really matter but:

```suggestion
      File.foreach(fn, chomp: true) do |line|
```

This may help if we encounter any windows line endings in the future.

Fryguy commented on bin/requirements.rb:
```diff
@@ -0,0 +1,187 @@
+    # some libraries list "python" instead of "python_version"
+    line.gsub!(/^python /, "python_version")
+    # change version 5+ to >=5
+    line.gsub!(/\([0-9.]*\)[+]/, '>=\1')
```
```suggestion
    line.gsub!(/\([0-9.]*\)\+/, '>=\1')
```

Fryguy commented on bin/requirements.rb:
```diff
@@ -0,0 +1,187 @@
+    # ignore package requirements for older version of pythons (assumption here)
+    return if line.match?(/python_version ?[=<]/)
+
+    # split on first space (or =)
+    # version can have multiple spaces
+    lib, ver = line.match(/([^ >=]*) ?(.*)/).captures
```
I'm not sure I understand this line, but I think this can more simply be

```suggestion
    # split on first space (or =)
    # version can have multiple spaces
    lib, ver = line.split(/[ =>]/, 2)
```

kbrock replied:
nice. Tested with a few and it works for most everything but this breaks down:

```
library==3.4
```

It looses the first `=` and turns into `library=3.4`

Maybe we want something like:
```
"library==22 ; <3".split(/([>= ]+)/, 2)
```

I'd then rewrite the comparison routine to keep the comparator separate, though then again, it does work the way it is.

Fryguy commented on bin/requirements.rb:
```diff
@@ -0,0 +1,187 @@
+  end
+
+  # return highest, lowest for version comparison
+  def version_compare(a, b)
```
Can we use Gem::Version object here to do the heavy lifting or are the requirements not quite the same?  e.g. something like:

```suggestion
  def version_compare(a, b)
    [Gem::Version.new(a), Gem::Version.new(b)].sort.reverse
  end
```
```

Fryguy replied:
or perhaps Gem::Requirement...

```ruby
a = Gem::Requirement.new(">=4.5.0")
a.requirements
# => [[">=", Gem::Version.new("4.5.0")]]

a = Gem::Requirement.new([">=4.5.0", ">=1.2.3"])
a.requirements
# => [[">=", Gem::Version.new("4.5.0")], [">=", Gem::Version.new("1.2.3")]]
```

kbrock replied:
Python and ruby seem very close here - https://pip.pypa.io/en/stable/reference/requirements-file-format/

I do like the idea of requirements instead of doing this ourselves.
But in the end of the day, there are only 2 instances where it is not a simple "merge no dependency with a dependency"

And I'm not sure that we want to invent a full requirements.txt parser in ruby to handle the `,` and `;` and all the wonderful `python_version` stuff.

Fryguy commented:
I have a handful of Ruby comments but they can all be follow ups if needed.

kbrock commented:
update:
- Added venv information for setup
- Fixed new_requirements name typo
- Fixed add_target to better support directories and missing files
- Better "legacy" requirements detection
- Using `IO.foreach` (`File.foreach` didn't work)
- Capitalization for comments
- Smaller `python_version` line detection
- `;python_version` detection moved to `ver` processing (some day would be nice to improve)
- Fixed version 5+ regex
- pulled out `split_lib_ver`. Wasn't able to fix regexp - punting for now
-

kbrock commented:
ok, so with our list of requirements, we only have 1 or 2 version comparisons to make.

I think having a very rudimentary system and displaying the work so a person can double check the work makes the most sense.

Fryguy commented:
@kbrock This closed a lot of Mend issues 🎉 , but seems to have opened a few more. See https://github.com/ManageIQ/manageiq-rpm_build/issues?q=is%3Aissue+sort%3Aupdated-desc+updated%3A2024-09-18T02%3A50%3A00-04%3A00..2024-09-19+author%3Aapp%2Fmend-bolt-for-github
